pyiron_base/job/worker.py

```diff
@@ -0,0 +1,129 @@
+        debug (bool): enable debug mode [True/False] (optional)
+    """
+    status, working_directory, job_id, _, submit_on_remote, debug = args
+    from pyiron_base.job.wrapper import JobWrapper
+    job_wrap = JobWrapper(
+        working_directory=working_directory,
+        job_id=job_id,
+        submit_on_remote=submit_on_remote,
+        debug=debug,
+    )
+    job_wrap.run()
+
+
+class WorkerJob(PythonTemplateJob):
+    """
+    The WorkerJob executes jobs linked to its master id.
+    """
+    def __init__(self, project, job_name):
+        super(WorkerJob, self).__init__(project, job_name)
+        self.input.project = None
+        self.input.cores_per_job = 1
+        self.input.sleep_interval = 10
+
+    @property
+    def project_to_watch(self):
+        rel_path = os.path.relpath(
+            self.input.project,
+            self.project.path
+        )
+        return self.project.open(rel_path)
+
+    @project_to_watch.setter
+    def project_to_watch(self, pr):
+        self.input.project = pr.path
+
+    @property
+    def cores_per_job(self):
+        return self.input.cores_per_job
+
+    @cores_per_job.setter
+    def cores_per_job(self, cores):
+        self.input.cores_per_job = int(cores)
+
+    @property
+    def sleep_interval(self):
+        return self.input.sleep_interval
+
+    @sleep_interval.setter
+    def sleep_interval(self, interval):
+        self.input.sleep_interval = int(interval)
+
+    # This function is executed
+    def run_static(self):
+        self.status.running = True
+        master_id = self.job_id
+        pr = self.project_to_watch
+        active_job_ids = []
+        with Pool(
+            processes=int(self.server.cores/self.cores_per_job),
+            initializer=worker_function,
+        ) as pool:
+            while True:
+                # Check the database if there are more calculation to execute
+                df = pr.job_table()
+                df_sub = df[
+                    (df["status"] == "submitted") &
+                    (df["masterid"] == master_id) &
+                    (~df["id"].isin(active_job_ids))
+                ]
+                if len(df_sub) > 0:  # Check if there are jobs to execute
+                    path_lst = [
+                        [pp, p, job_id]
+                        for pp, p, job_id in zip(
+                            df_sub["projectpath"].values,
+                            df_sub["project"].values,
+                            df_sub["id"].values
+                        ) if job_id not in active_job_ids]
+                    job_lst = [
+                        [True, p, job_id, None, False, False]
+                        if pp is None else
+                        [True, os.path.join(pp, p), job_id, None, False, False]
+                        for pp, p, job_id in path_lst
+                    ]
+                    active_job_ids += [j[2] for j in job_lst]
+                    _ = [pool.put(j) for j in job_lst]
```

Comment:
What's the reason not to use a plain `multiprocessing.Pool` & `map_async` here?  I'm a bit wary because I see that `aproc` uses `inspect.getsource` to transfer the function to the worker instead of pickle or dill.

Reply:
The list is growing while the worker is already executing it, so a simple map does not work. The current combination in the background uses a Multiprocessing pool plus a queue. In addition I also tried `dill` and `cloudpickle` but both failed, this might be related to the use of decorators.

Reply:
I don't understand why the growing list is a problem.

Inside the while loop the worker checks the job table for submitted jobs are assigned to it via the `master_id` and puts them in a list.  It could them `map_async` the list and mark the job ids as 'running' or 'queued' or whatever.  Then in the next iteration of the while loop there may be new submitted jobs which can be handled the same way or there may be submitted jobs that were already found in the last iteration, but since we marked them in the last iteration we can exclude them from the map in this iteration.

Reply:
The limitation of multiple `map_async` calls is that they can not be executed concurrently. So the second call to `map_async` is only executed once the first one is finished. The `async` only means you get control of the main process, it does not mean that the workers are asynchronous. That is why I created the `aproc` interface.

Reply:
From my testing this is not the case.  I can call `map_async` multiple times just fine and as long as there are available workers in the pool.  Once the pool is completely used you have to wait naturally, but I assume `aproc` is the same in this regard.

Here's my test script
```python
import multiprocessing as mp

def pow2(x):
    import time
    print(x)
    time.sleep(10)
    return x**2

with mp.Pool(4) as p:
    r1 = p.map_async(pow2, range(2))
    r2 = p.map_async(pow2, range(4, 8))

    r1.wait()
    print(r1.get())
    r2.wait()
    print(r2.get())
```
with output
```
0
1
4
5
<nothing happens here for 10s>
6
7
[0, 1]
[16, 25, 36, 49]
```

Reply:
I thought I tried that before but it is working now, so I reverted to `map_async` - thanks again for pointing me in this direction.